server/handlers/project_command_output_handler.go

```diff
@@ -83,32 +93,47 @@ func NewAsyncProjectCommandOutputHandler(
 		projectStatusUpdater:   projectStatusUpdater,
 		projectJobURLGenerator: projectJobURLGenerator,
 		projectOutputBuffers:   map[string][]string{},
+		pullToJobMapping:       map[models.PullContext]map[string]bool{},
 	}
 }
 
 func (p *AsyncProjectCommandOutputHandler) Send(ctx models.ProjectCommandContext, msg string) {
 	p.projectCmdOutput <- &models.ProjectCmdOutputLine{
-		ProjectInfo: ctx.PullInfo(),
-		Line:        msg,
+		JobID: ctx.JobID,
+		JobContext: models.JobContext{
+			HeadCommit: ctx.Pull.HeadCommit,
+			PullContext: models.PullContext{
+				PullNum:     ctx.Pull.Num,
+				Repo:        ctx.BaseRepo.Name,
+				ProjectName: ctx.ProjectName,
+				Workspace:   ctx.Workspace,
+			},
+		},
+		Line: msg,
 	}
 }
 
-func (p *AsyncProjectCommandOutputHandler) Register(projectInfo string, receiver chan string) {
-	p.addChan(receiver, projectInfo)
+func (p *AsyncProjectCommandOutputHandler) Register(jobID string, receiver chan string) {
+	p.addChan(receiver, jobID)
 }
 
 func (p *AsyncProjectCommandOutputHandler) Handle() {
 	for msg := range p.projectCmdOutput {
 		if msg.ClearBuffBefore {
-			p.clearLogLines(msg.ProjectInfo)
+			p.clearLogLines(msg.JobID)
+		} else {
+			if p.pullToJobMapping[msg.JobContext.PullContext] == nil {
+				p.pullToJobMapping[msg.JobContext.PullContext] = map[string]bool{}
+			}
+			p.pullToJobMapping[msg.JobContext.PullContext][msg.JobID] = true
```

Comment:
Is there a reason to use `PullContext` vs `JobContext` as a key here?

Reply:
Because a JobContext is specific to a pull at a certain commit, we cannot use JobContext to track all the jobs in a PR since a PR can have multiple revisions and thus multiple plan and apply jobs.

Reply:
As discussed during our one-on-one's, we will need this mapping for use-cases where log persistence is not configured. In that scenario, we'll have to wait for the PR to be merged before doing any clean up operations. So, we'll need the pullToJobMapping to keep track of all the jobs for a PR.